Clean up applyPropagate

tstat opened the pull request:
## Overview

Codebase updates are very slow for Stew's codebase. I observe ~76 seconds to execute Stew's update transcript on trunk (6bacb88e614a9883a107bcf50365f8a9025647b3). Profiling revealed that the bulk of the time was spent in `applyPropagate`'s `clearPropagated` call.  `clearPropagated` deletes the propagated metadata tags for all patch targets by looping over the patch targets and deleting them from the metadata Star3.

The call to `clearPropagated` appears to be unnecessary: `replaceFacts` traverses the intersection of the domain of (term/type)Edits and the corresponding metadata Star3 and properly assigns the propagated metadata values (see  `replaceTerms`/`replaceTypes`). So, `clearPropagated` would only be useful if it corrected metadata for references not in this intersection. If a reference is not in the Star3 then there is nothing to correct. If a reference is in the Star3 but not in the domain of `termEdits` then it would be erroneous to label it as not propagated.

I observe ~21 seconds to execute Stew's update transcript with this change.

mitchellwrosen commented on unison-cli/src/Unison/Codebase/Editor/Propagate.hs:
```diff
@@ -632,23 +632,32 @@ applyPropagate patch Edits {..} = do
+          updateMetadatas Just $
             Star3.replaceFacts replaceType typeEdits _types
 
+        projRef :: Referent -> Maybe Reference
```
Suggestion: `Referent.toTermReference`

tstat committed:
remove unnecessary call to Metadata.delete in `replaceTerm`

`replaceTerm` is bound to `apply` in `Star3.replaceFacts` which
supplies the `Star3` after first replacing all occurrences of `r` with
`r'` (see `replaceFact`)

So, there are no relations with a lhs of `r` in the `Star3`, and the
metadata delete will not have an effect.

tstat committed:
remove `clearPropagated` from `updateMetadatas`

The call to clearPropagated appears to be unnecessary: replaceFacts
traverses the intersection of the domain of (term/type)Edits and the
corresponding metadata Star3 and properly assigns the propagated
metadata values (see replaceTerms/replaceTypes). So, clearPropagated
would only have an effect if it corrected metadata for references not
in this intersection. If a reference is not in the Star3 then there is
nothing to correct. If a reference is in the Star3 but not in the
domain of termEdits then it would be erroneous to label it as not
propagated.

For example:

Suppose a codebase has only one term, B, that has the propagated
metadata. We apply a patch with the mapping A -> B. Star3.replaceFacts
has no effect since A is not in the codebase. The call to
`clearPropagated` will remove the propagated metadata from B
though. This doesn't seem correct though, B wasn't introduced by the
patch or a direct edit by the user.

tstat committed:
Don't use record wildcards in applyPropagate
